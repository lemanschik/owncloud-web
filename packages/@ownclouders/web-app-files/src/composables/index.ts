export * from './actions'
export * from './archiverService'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './spaces'
export * from './viewMode'
