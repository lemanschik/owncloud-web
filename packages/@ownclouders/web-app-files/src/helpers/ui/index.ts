export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
export * as resourceTiles from './resourceTiles'
