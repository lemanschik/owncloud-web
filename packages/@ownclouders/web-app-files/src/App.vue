<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="files-wrapper oc-width-expand oc-height-1-1 oc-flex-wrap" />
  </main>
</template>
<script lang="ts">
import { defineComponent, onBeforeUnmount, watch, ref } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    const dragareaEnabled = ref(false)
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const dragOver = eventBus.subscribe('drag-over', onDragOver)
    const dragOut = eventBus.subscribe('drag-out', hideDropzone)
    const drop = eventBus.subscribe('drop', hideDropzone)

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
    return { dragareaEnabled }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
