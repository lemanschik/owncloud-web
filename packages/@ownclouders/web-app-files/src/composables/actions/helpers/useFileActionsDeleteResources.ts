import { Store } from 'vuex'
import { cloneStateObject } from '../../../helpers/store'
import { isSameResource } from '../../../helpers/resource'
import { buildWebDavFilesTrashPath } from '../../../helpers/resources'
import { buildWebDavSpacesTrashPath, Resource, SpaceResource } from 'web-client/src/helpers'
import PQueue from 'p-queue'
import { isLocationTrashActive, isLocationSpacesActive } from '../../../router'
import { dirname } from 'path'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
import { computed, unref } from 'vue'
import {
  useCapabilitySpacesEnabled,
  useCapabilityShareJailEnabled,
  useClientService,
  useRouter,
  useStore,
  useLoadingService
} from 'web-pkg/src/composables'
import { useGettext } from 'vue3-gettext'
import { ref } from 'vue'

export const useFileActionsDeleteResources = ({ store }: { store?: Store<any> }) => {
  store = store || useStore()
  const router = useRouter()
  const language = useGettext()
  const { $gettext, $ngettext, interpolate: $gettextInterpolate } = language
  const hasShareJail = useCapabilityShareJailEnabled()
  const hasSpacesEnabled = useCapabilitySpacesEnabled()
  const clientService = useClientService()
  const loadingService = useLoadingService()
  const { owncloudSdk } = clientService

  const queue = new PQueue({ concurrency: 4 })
  const deleteOps = []
  const resourcesToDelete = ref([])

  const isInTrashbin = computed(() => {
    return isLocationTrashActive(router, 'files-trash-generic')
  })

  const resources = computed(() => {
    return cloneStateObject(unref(resourcesToDelete))
  })

  const dialogTitle = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'
    let title = null

    if (currentResources.length === 1) {
      if (isFolder) {
        title = unref(isInTrashbin)
          ? $gettext('Permanently delete folder %{name}')
          : $gettext('Delete folder %{name}')
      } else {
        title = unref(isInTrashbin)
          ? $gettext('Permanently delete file %{name}')
          : $gettext('Delete file %{name}')
      }
      return $gettextInterpolate(
        title,
        {
          name: currentResources[0].name
        },
        true
      )
    }

    title = unref(isInTrashbin)
      ? $ngettext(
          'Permanently delete selected resource?',
          'Permanently delete %{amount} selected resources?',
          currentResources.length
        )
      : $ngettext(
          'Delete selected resource?',
          'Delete %{amount} selected resources?',
          currentResources.length
        )

    return $gettextInterpolate(title, { amount: currentResources.length }, false)
  })

  const dialogMessage = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'

    if (currentResources.length === 1) {
      if (isFolder) {
        return unref(isInTrashbin)
          ? $gettext(
              'Are you sure you want to delete this folder? All its content will be permanently removed. This action cannot be undone.'
            )
          : $gettext('Are you sure you want to delete this folder?')
      }
      return unref(isInTrashbin)
        ? $gettext(
            'Are you sure you want to delete this file? All its content will be permanently removed. This action cannot be undone.'
          )
        : $gettext('Are you sure you want to delete this file?')
    }

    return unref(isInTrashbin)
      ? $gettext(
          'Are you sure you want to delete all selected resources? All their content will be permanently removed. This action cannot be undone.'
        )
      : $gettext('Are you sure you want to delete all selected resources?')
  })

  const trashbin_deleteOp = (space, resource) => {
    const path = unref(hasShareJail)
      ? buildWebDavSpacesTrashPath(space.id)
      : buildWebDavFilesTrashPath(store.getters.user.id)

    return owncloudSdk.fileTrash
      .clearTrashBin(path, resource.id)
      .then(() => {
        store.dispatch('Files/removeFilesFromTrashbin', [resource])
        const translated = $gettext('"%{file}" was deleted successfully')
        store.dispatch('showMessage', {
          title: $gettextInterpolate(translated, { file: resource.name }, true)
        })
      })
      .catch((error) => {
        if (error.statusCode === 423) {
          // TODO: we need a may retry option ....
          const p = queue.add(() => {
            return trashbin_deleteOp(space, resource)
          })
          deleteOps.push(p)
          return
        }

        console.error(error)
        const translated = $gettext('Failed to delete "%{file}"')
        store.dispatch('showMessage', {
          title: $gettextInterpolate(translated, { file: resource.name }, true),
          status: 'danger'
        })
      })
  }

  const trashbin_delete = (space: SpaceResource) => {
    // TODO: use clear all if all files are selected
    store.dispatch('toggleModalConfirmButton')
    // FIXME: Implement proper batch delete and add loading indicator
    for (const file of unref(resources)) {
      const p = queue.add(() => {
        return trashbin_deleteOp(space, file)
      })
      deleteOps.push(p)
    }

    Promise.all(deleteOps).then(() => {
      store.dispatch('hideModal')
      store.dispatch('toggleModalConfirmButton')
    })
  }

  const filesList_delete = (space: SpaceResource) => {
    return loadingService.addTask(
      (loadingCallbackArgs) => {
        return store
          .dispatch('Files/deleteFiles', {
            ...language,
            space,
            files: unref(resources),
            clientService,
            loadingCallbackArgs
          })
          .then(async () => {
            store.dispatch('hideModal')
            store.dispatch('toggleModalConfirmButton')

            // Load quota
            if (
              isLocationSpacesActive(router, 'files-spaces-generic') &&
              !['public', 'share'].includes(space?.driveType)
            ) {
              if (unref(hasSpacesEnabled)) {
                const graphClient = clientService.graphAuthenticated
                const driveResponse = await graphClient.drives.getDrive(
                  unref(resources)[0].storageId
                )
                store.commit('runtime/spaces/UPDATE_SPACE_FIELD', {
                  id: driveResponse.data.id,
                  field: 'spaceQuota',
                  value: driveResponse.data.quota
                })
              } else {
                const user = await owncloudSdk.users.getUser(store.getters.user.id)
                store.commit('SET_QUOTA', user.quota)
              }
            }

            if (
              unref(resourcesToDelete).length &&
              isSameResource(unref(resourcesToDelete)[0], store.getters['Files/currentFolder'])
            ) {
              return router.push(
                createFileRouteOptions(space, {
                  path: dirname(unref(resourcesToDelete)[0].path),
                  fileId: unref(resourcesToDelete)[0].parentFolderId
                })
              )
            }
          })
      },
      { indeterminate: false }
    )
  }

  const deleteHelper = (space: SpaceResource) => {
    store.dispatch('toggleModalConfirmButton')

    unref(isInTrashbin) ? trashbin_delete(space) : filesList_delete(space)
  }

  const displayDialog = (space: SpaceResource, resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    const modal = {
      variation: 'danger',
      title: unref(dialogTitle),
      message: unref(dialogMessage),
      cancelText: $gettext('Cancel'),
      confirmText: $gettext('Delete'),
      onCancel: () => {
        store.dispatch('hideModal')
      },
      onConfirm: () => {
        deleteHelper(space)
      }
    }

    store.dispatch('createModal', modal)
  }

  return {
    displayDialog,
    // HACK: exported for unit tests:
    filesList_delete
  }
}
