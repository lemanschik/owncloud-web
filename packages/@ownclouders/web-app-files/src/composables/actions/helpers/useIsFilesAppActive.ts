import { computed, unref } from 'vue'
import { activeApp, useRoute } from 'web-pkg/src/composables'

const isFilesAppActive = (activeApp: string): boolean => {
  // FIXME: we should use this constant but it somehow breaks the unit tests
  // return activeApp === FilesApp.appInfo.id
  return activeApp === 'files'
}

export const useIsFilesAppActive = () => {
  const currentRoute = useRoute()

  return computed(() => isFilesAppActive(activeApp(unref(currentRoute))))
}
