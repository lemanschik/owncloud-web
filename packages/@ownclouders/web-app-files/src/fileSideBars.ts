import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import SharesPanel from './components/SideBar/Shares/SharesPanel.vue'
import TagsPanel from './components/SideBar/TagsPanel.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetails.vue'
import SpaceNoSelection from 'web-pkg/src/components/sideBar/Spaces/SpaceNoSelection.vue'
import {
  isLocationTrashActive,
  isLocationPublicActive,
  isLocationSpacesActive,
  isLocationSharesActive
} from './router'
import { spaceRoleEditor, spaceRoleManager } from 'web-client/src/helpers/share'

import { Panel } from '../../web-pkg/src/components/sideBar'

import { Resource, User } from 'web-client'
import { Router } from 'vue-router'
import SpaceDetailsMultiple from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetailsMultiple.vue'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  resource,
  selectedFiles,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  resource: Resource
  selectedFiles: Resource[]
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ router, multipleSelection, rootFolder, resource }): Panel => ({
    app: 'no-selection',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return (
        !multipleSelection &&
        !isLocationSpacesActive(router, 'files-spaces-projects') &&
        (!resource || (rootFolder && resource?.type !== 'space'))
      )
    }
  }),
  ({ router, resource, multipleSelection }): Panel => ({
    app: 'no-selection',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceNoSelection,
    default: () => true,
    get enabled() {
      return (
        isLocationSpacesActive(router, 'files-spaces-projects') && !multipleSelection && !resource
      )
    }
  }),
  ({ router, multipleSelection, rootFolder, resource }) => ({
    app: 'details',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default: !isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return !!(
        !isLocationTrashActive(router, 'files-trash-generic') &&
        !multipleSelection &&
        !rootFolder &&
        resource &&
        resource.type !== 'space'
      )
    }
  }),
  ({ multipleSelection, router }) => ({
    app: 'details-multiple',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    componentAttrs: {
      get showSpaceCount() {
        return (
          !isLocationSpacesActive(router, 'files-spaces-generic') &&
          !isLocationSharesActive(router, 'files-shares-with-me') &&
          !isLocationTrashActive(router, 'files-trash-generic')
        )
      }
    },
    default: () => true,
    get enabled() {
      return multipleSelection && !isLocationSpacesActive(router, 'files-spaces-projects')
    }
  }),
  ({ multipleSelection, selectedFiles, router }) => ({
    app: 'details-space-multiple',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetailsMultiple,
    componentAttrs: {
      selectedSpaces: selectedFiles
    },
    default: () => true,
    get enabled() {
      return multipleSelection && isLocationSpacesActive(router, 'files-spaces-projects')
    }
  }),
  ({ multipleSelection, resource }) => ({
    app: 'details-space',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: () => true,
    get enabled() {
      return resource?.type === 'space' && !multipleSelection
    }
  }),
  ({ router, multipleSelection, rootFolder, resource }) => ({
    app: 'actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default: isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return !!(!multipleSelection && !rootFolder && resource && resource.type !== 'space')
    }
  }),
  ({ multipleSelection, resource, user }) => ({
    app: 'space-actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (multipleSelection) {
        return false
      }
      if (resource?.type !== 'space') {
        return false
      }
      return !![
        ...resource.spaceRoles[spaceRoleManager.name],
        ...resource.spaceRoles[spaceRoleEditor.name]
      ].find((role) => role.id === user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, resource }) => ({
    app: 'sharing',
    icon: 'user-add',
    iconFillType: 'line',
    title: $gettext('Shares'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: false,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      if (multipleSelection || rootFolder || !resource || resource.type === 'space') {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ capabilities, resource, router, multipleSelection, rootFolder }) => ({
    app: 'tags',
    icon: 'price-tag-3',
    iconFillType: 'line',
    title: $gettext('Tags'),
    component: TagsPanel,
    componentAttrs: {},
    get enabled() {
      if (
        !capabilities?.files?.tags ||
        multipleSelection ||
        rootFolder ||
        !resource ||
        resource.type === 'space'
      ) {
        return false
      }
      if (typeof resource.canEditTags !== 'function' || !resource.canEditTags()) {
        return false
      }
      return !(
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      )
    }
  }),
  ({ multipleSelection, resource, capabilities }) => ({
    app: 'space-share',
    icon: 'group',
    title: $gettext('Members'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: true,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      return resource?.type === 'space' && !multipleSelection
    }
  }),
  ({ capabilities, resource, router, multipleSelection, rootFolder }) => ({
    app: 'versions',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }
      return (
        !!capabilities.core && resource && resource.type !== 'folder' && resource.type !== 'space'
      )
    }
  })
]

export default panelGenerators
