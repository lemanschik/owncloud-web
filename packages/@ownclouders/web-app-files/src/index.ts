import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { Registry, ArchiverService } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'
import { AppNavigationItem, AppReadyHookArgs } from 'web-pkg/src/apps'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import { isPersonalSpaceResource } from 'web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find((drive) =>
          isPersonalSpaceResource(drive)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0'
      }
    }
  ]
}

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    },
    Trash: {
      Overview: TrashOverview
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store, globalProperties }: AppReadyHookArgs) {
    const { $clientService } = globalProperties
    Registry.sdkSearch = new SDKSearch(store, router, $clientService)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)
    globalProperties.$archiverService = new ArchiverService(
      $clientService,
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
  }
}
