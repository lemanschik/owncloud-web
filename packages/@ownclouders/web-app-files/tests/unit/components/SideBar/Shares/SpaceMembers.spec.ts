import SpaceMembers from 'web-app-files/src/components/SideBar/Shares/SpaceMembers.vue'
import {
  spaceRoleManager,
  spaceRoleViewer,
  ShareTypes,
  spaceRoleEditor,
  Share
} from 'web-client/src/helpers/share'
import { mock } from 'jest-mock-extended'
import { ProjectSpaceResource, SpaceResource, User } from 'web-client/src/helpers'
import {
  createStore,
  defaultPlugins,
  mount,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks,
  RouteLocation
} from 'web-test-helpers'

const memberMocks = {
  [spaceRoleManager.name]: {
    id: '1',
    shareType: ShareTypes.spaceUser.value,
    collaborator: {
      name: 'alice',
      displayName: 'alice'
    },
    role: {
      name: spaceRoleManager.name
    }
  },
  [spaceRoleEditor.name]: {
    id: '2',
    shareType: ShareTypes.spaceUser.value,
    collaborator: {
      onPremisesSamAccountName: 'Einstein',
      displayName: 'einstein'
    },
    role: {
      name: spaceRoleEditor.name
    }
  },
  [spaceRoleViewer.name]: {
    id: '3',
    shareType: ShareTypes.spaceUser.value,
    collaborator: {
      onPremisesSamAccountName: 'Marie',
      displayName: 'marie'
    },
    role: {
      name: spaceRoleViewer.name
    }
  }
}

describe('SpaceMembers', () => {
  describe('invite collaborator form', () => {
    it('renders the form when the current user is a manager of an enabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => true, disabled: false })
      const wrapper = getWrapper({ space })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
    it('does not render the form when the current user is no manager of an enabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => false, disabled: false })
      const wrapper = getWrapper({ space })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
    it('does not render the form when the current user is a manager of a disabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => true, disabled: true })
      const wrapper = getWrapper({ space })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
  })

  describe('existing members', () => {
    it('can edit when current user is manager of an enabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => true, disabled: false })
      const wrapper = getWrapper({ space })
      expect(
        wrapper.findAllComponents<any>('collaborator-list-item-stub').at(1).props().modifiable
      ).toEqual(true)
    })
    it('can not edit when current user is not a manager of an enabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => false, disabled: false })
      const wrapper = getWrapper({ space })
      expect(
        wrapper.findAllComponents<any>('collaborator-list-item-stub').at(1).props().modifiable
      ).toEqual(false)
    })
    it('can not edit when current user is manager of a disabled space', () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => true, disabled: true })
      const wrapper = getWrapper({ space })
      expect(
        wrapper.findAllComponents<any>('collaborator-list-item-stub').at(1).props().modifiable
      ).toEqual(false)
    })
  })

  describe('deleting members', () => {
    it('reacts on delete events by collaborator list items', async () => {
      const spyOnCollaboratorDeleteTrigger = jest.spyOn(
        (SpaceMembers as any).methods,
        '$_ocCollaborators_deleteShare_trigger'
      )

      const user = mock<User>({ id: memberMocks.manager.collaborator.name })
      const wrapper = getWrapper({ user })
      ;(wrapper.findComponent<any>('collaborator-list-item-stub').vm as any).$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(spyOnCollaboratorDeleteTrigger).toHaveBeenCalledTimes(1)
    })
    it('calls "deleteSpaceMember" when successful', async () => {
      const wrapper = getWrapper()
      const deleteSpaceMemberSpy = jest.spyOn(wrapper.vm, 'deleteSpaceMember')
      const share = mock<Share>()
      await wrapper.vm.$_ocCollaborators_deleteShare(share)
      expect(deleteSpaceMemberSpy).toHaveBeenCalled()
    })
    it('shows a message when an error occurs', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)
      const wrapper = getWrapper()
      jest.spyOn(wrapper.vm, 'deleteSpaceMember').mockRejectedValue(new Error())
      const showMessageSpy = jest.spyOn(wrapper.vm, 'showMessage')
      const share = mock<Share>()
      await wrapper.vm.$_ocCollaborators_deleteShare(share)
      expect(showMessageSpy).toHaveBeenCalled()
    })
    it('redirects to the "files-spaces-projects"-page when the current user has been removed', async () => {
      const user = mock<User>({ id: memberMocks.manager.collaborator.name })
      const wrapper = getWrapper({ user })
      jest.spyOn(wrapper.vm, 'deleteSpaceMember')
      await wrapper.vm.$_ocCollaborators_deleteShare(memberMocks.manager)
      expect(wrapper.vm.$router.push).toHaveBeenCalled()
    })
    it('refreshes the page when the current user has been removed on the "files-spaces-projects"-page', async () => {
      const user = mock<User>({ id: memberMocks.manager.collaborator.name })
      const wrapper = getWrapper({ user, currentRouteName: 'files-spaces-projects' })
      jest.spyOn(wrapper.vm, 'deleteSpaceMember')
      await wrapper.vm.$_ocCollaborators_deleteShare(memberMocks.manager)
      expect(wrapper.vm.$router.go).toHaveBeenCalled()
    })
  })

  describe('filter', () => {
    it('toggles the filter on click', async () => {
      const space = mock<ProjectSpaceResource>({ isManager: () => true })
      const wrapper = getWrapper({ mountType: mount, space })
      expect(wrapper.vm.isFilterOpen).toBeFalsy()
      await wrapper.find('.open-filter-btn').trigger('click')
      expect(wrapper.vm.isFilterOpen).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getWrapper({
  mountType = shallowMount,
  space = mock<SpaceResource>(),
  spaceMembers = [memberMocks.manager, memberMocks.editor, memberMocks.viewer],
  user = mock<User>(),
  currentRouteName = 'files-spaces-generic'
} = {}) {
  const storeOptions = {
    ...defaultStoreMockOptions,
    state: { user },
    getters: {
      ...defaultStoreMockOptions.getters,
      user: () => user,
      configuration: jest.fn(() => ({
        options: { contextHelpers: true, sidebar: { shares: { showAllOnLoad: true } } }
      }))
    }
  }
  storeOptions.modules.runtime.modules.spaces.getters.spaceMembers.mockImplementation(
    () => spaceMembers
  )
  const store = createStore(storeOptions)
  return mountType(SpaceMembers, {
    global: {
      plugins: [...defaultPlugins(), store],
      mocks: defaultComponentMocks({
        currentRoute: mock<RouteLocation>({ name: currentRouteName })
      }),
      provide: {
        space,
        resource: space
      },
      stubs: {
        OcButton: false,
        'oc-icon': true,
        'oc-spinner': true,
        'avatar-image': true,
        'role-dropdown': true,
        'edit-dropdown': true,
        'invite-collaborator-form': true,
        'collaborator-list-item': true
      }
    }
  })
}
