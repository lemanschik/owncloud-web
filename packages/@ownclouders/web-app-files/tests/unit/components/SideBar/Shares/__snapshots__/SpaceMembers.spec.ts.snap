// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceMembers filter toggles the filter on click 1`] = `
<div class="oc-position-relative" id="oc-files-sharing-sidebar">
  <div class="oc-flex oc-flex-between oc-flex-center oc-flex-middle">
    <div class="oc-flex oc-py-s">
      <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-current-language="en" data-msgid="Members">Members</h3>
      <div class="oc-contextual-helper oc-pl-xs">
        <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="oc-contextual-helper-1-button" type="button">
          <!-- @slot Content of the button -->
          <oc-icon-stub accessiblelabel="" color="" filltype="line" name="question" size="small" type="span" variation="passive"></oc-icon-stub>
        </button>
        <div class="oc-drop oc-box-shadow-medium oc-rounded oc-width-1-1 oc-info-drop" id="oc-contextual-helper-1">
          <div class="oc-card oc-card-body oc-background-secondary oc-p-m">
            <div class="info-drop-content">
              <div class="oc-flex oc-flex-between info-header oc-border-b oc-pb-s">
                <span class="oc-m-rm info-title">Add members to this Space</span>
                <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" type="button">
                  <!-- @slot Content of the button -->
                  <oc-icon-stub accessiblelabel="" color="" filltype="line" name="close" size="medium" type="span" variation="inherit"></oc-icon-stub>
                </button>
              </div>
              <p class="info-text">Enter a name to add people or groups as members to this Space.</p>
              <dl class="info-list">
                <dt>Member capabilities</dt>
                <dd>Members are able to see who has access to this space and access all files in this space. Read or write permissions can be set by assigning a role.</dd>
                <dt>Space manager capabilities</dt>
                <dd>Members with the Manager role are able to edit all properties and content of a Space, such as adding or removing members, sharing subfolders with non-members, or creating links to share.</dd>
              </dl>
              <!--v-if-->
              <a class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw info-more-link" href="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" target="_blank">Read more</a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="oc-flex">
      <div class="oc-flex">
        <div class="oc-text-truncate space-members-filter oc-mr-s">
          <label class="oc-label" for="oc-textinput-2"></label>
          <div class="oc-position-relative">
            <!--v-if-->
            <input aria-invalid="false" class="oc-text-input oc-input oc-rounded" id="oc-textinput-2" placeholder="Filter members" type="text">
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
        <button aria-label="Close filter" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw close-filter-btn" type="button">
          <!-- @slot Content of the button -->
          <oc-icon-stub fill-type="line" name="close" size="small"></oc-icon-stub>
        </button>
      </div>
    </div>
  </div>
  <!--v-if-->
  <ul aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm" id="files-collaborators-list">
    <li>
      <collaborator-list-item-stub deniable="false" issharedenied="false" modifiable="false" resourcename="" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub deniable="false" issharedenied="false" modifiable="false" resourcename="" share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub deniable="false" issharedenied="false" modifiable="false" resourcename="" share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
</div>
`;
